[SPARK-44508][PYTHON][DOCS] Add user guide for Python user-defined table functions

allisonwang-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR adds a user guide for Python user-defined table functions (UDTFs) introduced in Spark 3.5.
<img width="468" alt="Screenshot 2023-08-04 at 14 46 13" src="https://github.com/apache/spark/assets/66282705/11f5dc5e-681b-4677-a466-1a23c0b8dd01">


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
To help users write Python UDTFs.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
docs test

allisonwang-db committed:
docs

itholic commented on examples/src/main/python/sql/udtf.py:
```diff
@@ -0,0 +1,169 @@
+  ./bin/spark-submit examples/src/main/python/sql/udtf.py
+"""
+
+# NOTE that this file is imported in user guide in PySpark documentation.
```
nit: "user guide" -> "User Guides" to follow official documentation name?

Also maybe adding a doc link(https://spark.apache.org/docs/latest/api/python/user_guide/index.html) would helpful?

allisonwang-db replied:
Yup it's on the user guide page. I will add a screenshot in the PR description.

itholic commented on examples/src/main/python/sql/udtf.py:
```diff
@@ -0,0 +1,169 @@
+            self.count += 1
+    
+        def terminate(self):
+            yield self.count,
```
qq: should we always yield the data as `tuple` for UDTF?

allisonwang-db replied:
Yes, each element corresponds to one column in the output schema.

itholic commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+Python User-defined Table Functions (UDTFs)
+===========================================
+
+Spark 3.5 introduces a new type of user-defined fucntion: Python user-defined table functions (UDTFs),
```
typo: "fucntion" -> "function"

itholic commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+
+            Yields:
+                tuple: A tuple representing a single row in the UDTF result relation.
+                       Yield thisas many times as needed to produce multiple rows.
```
typo?: "thisas" -> "this as"

itholic commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+
+            This method is required to implement.
+
+            Args:
```
I'm not pretty sure if we should follow [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#sections) style here since we're following them in overall PySpark code base. WDYT @HyukjinKwon ?

HyukjinKwon replied:
yeah should follow numpy doc style I think

itholic reviewed (commented):
Otherwise looks pretty fine to me

itholic commented:
I think attaching screen capture(or something visible stuffs for built documentation) in the PR description would be great!

allisonwang-db commented:
cc @dtenedor @ueshin

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """"
+            Evaluate the function using the given input arguments.
```
Can we also describe the mapping between provided scalar argument(s) and the attributes of any provided relation argument with the ordering of these `*args` here? For example, we could mention that each provided scalar argument with index i in 0...N gets assigned to `*args` index i here, with the exception of at most one input relation argument at provided index j, and its attributes are injected one-by-one into the `*args` starting at index j?

allisonwang-db replied:
I am thinking about this too, but I found it difficult to explain in words. The interface is the same as scalar UDFs so I think Spark users should be able to figure it out. I can provide more examples.

dtenedor replied:
👍 more examples should be helpful. Maybe we could also add:

```
The arguments provided to the UDTF call map to the values in this *args list
in order. Each provided scalar expression maps to exactly one value in this
*args list. Each provided TABLE argument of N columns maps to exactly N
values in this *args list, in order of the columns as they appear in the
table.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+                  or use "return" to produce multiple rows for the UDTF result relation at once.
+
+            Example:
+                def eval(self, x: int, y: int):
```
```suggestion
                def eval(self, x: int, y: int) -> Iterator[Any]:
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+
+            Example:
+                def eval(self, x: int, y: int):
+                    yield x + y, x - y
```
can we also add an example with a combination of scalar constant arguments and a relation input argument, to show how the mapping  from provided SQL arguments to the python `*args` works? Could we include a SQL query and its results with each example as well?

allisonwang-db replied:
Sure! I will add a simple one here and a more complex one in the example section below.

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+
+        def terminate(self) -> Iterator[Any]:
+            """
+            Called when the UDTF has processed all rows in a partition.
```
We haven't really precisely defined what comprises a partition yet. Should we define it using the definitions from https://github.com/apache/spark/pull/42100 and https://github.com/apache/spark/pull/42174? Alternatively if these docs are targeting Spark 3.5 but those PRs are only going into `master`, we could simply define a partition here as either (1) just one `eval` call with the provided scalar argument(s), if any, or (2) several `eval` calls with an undefined subset of the rows from the input relation. Then we can expand it later.

dtenedor reviewed (commented):
These docs look great, thanks Allison for working on this!

ueshin commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+            Initialize the user-defined table function (UDTF).
+
+            This method is optional to implement and is called once when the UDTF is
+            instantiated. Use it to perform any initialization required for the UDTF.
```
Also should mention that it should be a default constructor which doesn’t accept any extra arguments?

ueshin commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+                - The result must be a tuple.
+                - UDTFs do not accept keyword arguments on the calling side.
+                - Use "yield" to produce one row at a time for the UDTF result relation,
+                  or use "return" to produce multiple rows for the UDTF result relation at once.
```
We might want an example with `return`?

ueshin commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,140 @@
+Apache Arrow
+------------
+Apache Arrow is an in-memory columnar data format that is used in Spark to efficiently transfer
+data between JVM and Python processes. Apache Arrow is by default enabled for Python UDTFs.
```
It will be disabled by default in 3.5. #42329

allisonwang-db replied:
Yup I will update this doc

allisonwang-db committed:
address comments

allisonwang-db committed:
more updates

dstrodtman-db commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,203 @@
+            -----
+            - The result of the function must be a tuple representing a single row
+              in the UDTF result relation.
+            - UDTFs do not ccept keyword arguments during the function call.
```
```suggestion
            - UDTFs do not accept keyword arguments during the function call.
```

allisonwang-db commented:
@ueshin @dtenedor @itholic @allanf-db @dstrodtman-db I've addressed the comments; PTAL thanks!

allisonwang-db committed:
address comments

allisonwang-db commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,222 @@
+Advanced Features
+-----------------
+
+TABLE input argument
+~~~~~~~~~~~~~~~~~~~~
+Python UDTFs can also take a TABLE as input argument, and it can be used in conjunction 
+with scalar input arguments.
+By default, you are allowed to have only one TABLE argument as input, primarily for 
+performance reasons. If you need to have more than one TABLE input argument, 
+you can enable this by setting ``spark.sql.tvf.allowMultipleTableArguments.enabled`` to ``true``.
+
+.. literalinclude:: ../../../../../examples/src/main/python/sql/udtf.py
+    :language: python
+    :lines: 178-197
+    :dedent: 4
```
@ueshin @dtenedor I've added this TABLE argument support under the advanced feature section. We should add more examples here and in the udtf function docstring.

allisonwang-db replied:
Created [SPARK-44746](https://issues.apache.org/jira/browse/SPARK-44746)

allisonwang-db commented on python/pyspark/sql/functions.py:
```diff
@@ -15639,13 +15639,6 @@ def udtf(
     Use "yield" to produce one row for the UDTF result relation as many times
     as needed. In the context of a lateral join, each such result row will be
     associated with the most recent input row consumed from the "eval" method.
-    Or, use "return" to produce multiple rows for the UDTF result relation at
```
We recommend always using the `yield` statement.

johnayoub commented:
@allisonwang-db would we able to use this feature to return a dataframe? I think this will be extremely useful especially that some functions such as dropDuplicates have no equivalent in SQL and wrapping them would be helpful.

allisonwang-db commented:
@johnayoub A Python UDTF is a table-valued function and it returns a dataframe. However, I don't think you can directly use dataframe functions like dropDuplicates directly inside the UDTF.

dstrodtman-db commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,222 @@
+            
+            The arguments provided to the UDTF call are mapped to the values in the
+            `*args` list sequentially. Each provided scalar expression maps to exactly
+            one value in this `*args` list. Each provided TABLE argument of N columns
```
@dtenedor Here's the line

dtenedor replied:
@allisonwang-db it turns out this part about `TABLE` arguments is wrong (I think I suggested it before, sorry). Instead of:

```
Each provided TABLE argument of N columns
maps to exactly N values in this `*args` list, in the order of the columns
as they appear in the table.
```

it should be something like

```
Each provided TABLE argument maps to a pyspark.sql.Row object containing
the columns in the order they appear in the provided input relation.
```

allisonwang-db replied:
Updated!

allisonwang-db committed:
address comments and update examples

ueshin commented on examples/src/main/python/sql/udtf.py:
```diff
@@ -0,0 +1,230 @@
+    # Invoke the UDTF in PySpark using the SquareNumbers class directly.
+    SquareNumbers(lit(1), lit(3)).show()  # type: ignore
+    # +---+------+
+    # |num|squred|
```
ditto.

ueshin commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,215 @@
+
+.. literalinclude:: ../../../../../examples/src/main/python/sql/udtf.py
+    :language: python
+    :lines: 38-42
```
We need to revisit the line numbers? I guess this is 36-40.
And ditto for the following examples.

allisonwang-db committed:
remove mypy checks

allisonwang-db committed:
address comments

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+===========================================
+
+Spark 3.5 introduces Python user-defined table functions (UDTFs), a new type of user-defined function. 
+Unlike scalar functions that return a single result value for every input, UDTFs is invoked in the ``FROM``
```
```suggestion
Unlike scalar functions that return a single result value from each call, each UDTF is invoked in the ``FROM``
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+.. currentmodule:: pyspark.sql.functions
+
+To implement a Python UDTF, you can define a class implementing the methods:
```
```suggestion
To implement a Python UDTF, you can define a class implementing the following methods:
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
```
```suggestion
            Initializes the user-defined table function (UDTF). This is optional.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+            UDTF is instantiated on the executor side.
+            
+            Any class fields assigned in this method will be available for subsequent
+            calls to the `eval` and `terminate` methods.
```
```suggestion
            calls to the `eval` and `terminate` methods. This class instance
            will remain alive until all rows in the current partition have
            been consumed by the `eval` method.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+            Notes
+            -----
+            - This method does not accept any extra arguments.
```
```suggestion
            - This method does not accept any extra arguments. Only the default
              constructor is supported.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """
+            Evaluate the function using the given input arguments.
```
```suggestion
            Evaluates the function using the given input arguments.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+            Argument Mapping:
+            - Each provided scalar expression maps to exactly one value in the
+              `*args` list.
```
```suggestion
              `*args` list. Or, if the `eval` method accepts `**kwargs`, these
              provide the corresponding argument names if the UDTF call passed
              SQL named arguments.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+            - Each provided scalar expression maps to exactly one value in the
+              `*args` list.
+            - Each provided table argument maps to a pyspark.sql.Row object containing
+              the columns in the order they appear in the provided input relation.
```
```suggestion
              the columns in the order they appear in the provided input relation,
              and with the names computed by the query analyzer.
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+            -----
+            - The result of the function must be a tuple representing a single row
+              in the UDTF result relation.
+            - UDTFs currently do not accept keyword arguments during the function call.
```
this is now obsolete since @ueshin  implemented this functionality.

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+            Examples
+            --------
+            >>> def eval(self, x: int, y: int) -> Iterator[Any]:
```
This example does not use `*args` but accepts the exact arguments as a list instead :) can we mention this is possible in the above description as well, and add an example with `*args*` and `**kwargs` as well, since all three ways are supported?

allisonwang-db replied:
I was thinking about this, but then I realized that using `*args` is not as readable as using variable names. It might not be something we should encourage users to use. For example:
```
def eval(self, *args):
    yield args[0], args[1]
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+            This method is optional to implement and is useful for performing any
+            cleanup or finalization operations after the UDTF has finished processing
+            all rows. It can also be used to yield additional rows if needed.
```
```suggestion
            all rows. It can also be used to yield additional rows if needed.
            Table functions that consume all rows in the entire input partition
            and then compute and return the entire output table can do so from
            this method as well (please be mindful of memory usage when doing
            this).
```

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+            ------
+            tuple
+                A tuple representing a single row in the UDTF result relation.
+                Yield this if you want to return additional rows during termination.
```
should we mention here the tricky detail that you have to include a trailing comma when yielding a row of just one value (here and above in the `eval` description)?

allisonwang-db replied:
Yea let me add an example

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+
+
+The return type of the UDTF defines the schema of the table it outputs. 
+It must be either a ``StructType`` or a DDL string representing a struct type.
```
should we put an example with this DDL string as well? It looks useful :)

allisonwang-db replied:
Will do. All the examples below are actually using DDL strings, but I couldn't find any documentation on this. cc @HyukjinKwon do you know if we have documentation on DDL strings of pyspark types?

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+Advanced Featuress
+------------------
+
+TABLE input argument
```
I would recommend to propose this as the primary way of passing relation arguments, rather than in the "additional features" section, since this syntax conforms to the SQL standard.

One way is to just swap the LATERAL syntax to this "advanced features" section instead.

allisonwang-db replied:
Moved. But we might need to improve it in the future ([SPARK-44746](https://issues.apache.org/jira/browse/SPARK-44746))

dtenedor commented on python/docs/source/user_guide/sql/python_udtf.rst:
```diff
@@ -0,0 +1,216 @@
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
```
Should we document the polymorphic `analyze` method here as well?

dtenedor commented on examples/src/main/python/sql/udtf.py:
```diff
@@ -0,0 +1,230 @@
+
+    spark.udtf.register("filter_udtf", FilterUDTF)
+
+    spark.sql("SELECT * FROM filter_udtf(TABLE(SELECT * FROM range(10)))").show()
```
this is good, let's also add an example just passing a table by name directly as well, e.g. `TABLE(t)`?

allisonwang-db replied:
We can follow up in [SPARK-44746](https://issues.apache.org/jira/browse/SPARK-44746)

dtenedor reviewed (commented):
Thanks @allisonwang-db for putting in the work to get this drafted, the documentation will be very useful for Spark users!

allisonwang-db committed:
address comments

allisonwang-db committed:
retrigger build

allisonwang-db commented:
I've addressed all comments. cc @HyukjinKwon we should merge this soon in spark 3.5. I can create follow up PRs if there are additional comments. Thanks!

HyukjinKwon commented:
Merged to master and branch-3.5.
